File: Old Code/WurmRancher/WurmRancher/FeederLevel.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Collections.Generic;

namespace WurmRacher
{
    public class FeederLevel : TimedLevel
    {
        public FeederLevel(Theme theme) : base(theme, 80) { }

        public override string Name
        {
            get
            {
                return "Feeder Raiser";
            }
        }

        public override string Description
        {
            get
            {
                return string.Format("It looks like the wurms aren't ready yet... but you might as well get the feeders ready.  Get {0} feeders up to {1} to win!", maxed_feeders_to_win, Feeder.max_size);
            }
        }

        public override string QuickObjectives
        {
            get
            {
                return string.Format("Get {0} feeders to size {1}.", maxed_feeders_to_win, Feeder.max_size);
            }
        }
        public override string HighScoreName
        {
            get
            {
                return "FeederLevel";
            }
        }

        const int maxed_feeders_to_win = 8;
        Counter FeederCounter;
        List<Feeder> feeders;
        public override void InitializeLevel(GameControl theControl)
        {
            base.InitializeLevel(theControl);
            FeederCounter = new Counter("Feeders");
            FeederCounter.Value = 0;
            theControl.AddCounter(FeederCounter);
            theControl.GrowRandomPoisonWeed();
            theControl.GrowRandomPoisonWeed();
            theControl.GrowRandomWeed();
            theControl.GrowRandomWeed();
            feeders = new List<Feeder>();
            
            for (int i = 0; i < 3; i++)
            {
                AddFeeder(theControl);
            }                
            
            theControl.AddCreature(new GrassEater(theControl));
        }

        public override void Update(GameControl theControl)
        {
            base.Update(theControl);
            if (!GameOver)
            {
                if (IntervalTimeIsUp(4))
                    theControl.AddCreature(new GrassEater(theControl));
                if (IntervalTimeIsUp(3))
                    AddFeeder(theControl);
                if (IntervalTimeIsUp(9))
                    theControl.AddCreature(new Monster(theControl));
                if (IntervalTimeIsUp(6))
                    theControl.GrowRandomPoisonWeed();
                if (IntervalTimeIsUp(70, 25))
                    theControl.AddCreature(new BigMonster(theControl));
                if (IntervalTimeIsUp(7))
                    theControl.GrowRandomWeed();
            }
        }

        void AddFeeder(GameControl theControl)
        {
            Feeder f = new Feeder(theControl);
            f.EatsGrass += new EventHandler<GameEventArgs>(Feeder_eats);
            f.Removed += new EventHandler<GameEventArgs>(Feeder_removed);
            feeders.Add(f);
            theControl.AddCreature(f);
        }

        void Feeder_removed(object sender, GameEventArgs e)
        {
            feeders.Remove((Feeder)sender);
            UpdateCounter(e.theControl);
        }

        void UpdateCounter(GameControl theControl)
        {
            int count10 = 0;
            foreach (Feeder f in feeders)
                if (f.FeederSize >= Feeder.max_size)
                    count10++;
            FeederCounter.Value = count10;
            if (count10 >= maxed_feeders_to_win)
                Victory(theControl);
        }

        void Feeder_eats(object sender, GameEventArgs e)
        {
            UpdateCounter(e.theControl);            
        }



    }
}

File: ts/feederLevel.ts
import { BigMonster } from "./bigMonster.js";
import { Counter } from "./counter.js";
import { Feeder, max_fattened } from "./feeder.js";
import { AddCounter, AddCreature, AddCreatureOnEdge, GrowRandomPoisonWeed, GrowRandomWeed } from "./gameControl.js";
import { GrassEater } from "./grassEater.js";
import { Monster } from "./monster.js";
import { Theme } from "./theme.js";
import { TimedLevel } from "./timedLevel.js";

export class FeederLevel extends TimedLevel{
    maxed_feeders_to_win = 8;
    Name = "Feeder Raiser";
    Description = `It looks like the wurms aren't ready yet... but you might as well get the feeders ready.  Get ${this.maxed_feeders_to_win} feeders up to ${max_fattened} to win!`;
    QuickObjectives = `Get ${this.maxed_feeders_to_win} feeders to size ${max_fattened}.`;

    FeederCounter : Counter;

    feeders : Array<Feeder>;

    constructor(theme:Theme){
        super(theme,80000);
    }

    InitializeLevel(): void {
        super.InitializeLevel();
        this.FeederCounter = new Counter("Feeders");
        this.FeederCounter.Value = 0;
        AddCounter(this.FeederCounter);
        GrowRandomPoisonWeed();
        GrowRandomPoisonWeed();
        GrowRandomWeed();
        GrowRandomWeed();
        this.feeders = new Array<Feeder>();
        
        for (let i = 0; i < 3; i++) this.AddFeeder();

        AddCreatureOnEdge(new GrassEater());
    }

    Update(time_step: number): void {
        super.Update(time_step);
        if (this.IntervalTimeIsUp(4000))
            AddCreatureOnEdge(new GrassEater());
        if (this.IntervalTimeIsUp(3000))
            this.AddFeeder();
        if (this.IntervalTimeIsUp(9000))
            AddCreatureOnEdge(new Monster());
        if (this.IntervalTimeIsUp(6000))
            GrowRandomPoisonWeed();
        if (this.IntervalTimeIsUp(70000, 25000))
            AddCreatureOnEdge(new BigMonster());
        if (this.IntervalTimeIsUp(7000))
            GrowRandomWeed();

        if (this.IntervalTimeIsUp(200)){
            let fat_feeders_count = 0;
            for (let f of this.feeders){
                if (!(f.eaten) && f.fattened == max_fattened)
                    fat_feeders_count ++;
            }
            this.FeederCounter.Value = fat_feeders_count;
            if (fat_feeders_count >= this.maxed_feeders_to_win)
                this.Victory();
        }
    }

    AddFeeder(){
        let f = new Feeder();
        this.feeders.push(f);
        AddCreatureOnEdge(f);
    }


}